### PR TITLE
Use Microsoft.VisualStudio.Threading to resolve deadlocks 

### PR DESCRIPTION
**Pull request by A9G-Data-Droid**
All tests are passing. I still need to integrate this into my main project and do a prove out there but I wanted to get this PR in front of you so I can get feedback on what I have attempted here.

I apologize for the large PR. As you know, async is contagious and must be implemented across a full chain of events.

I originally set out to remove `Monitor.Wait` from `QuartzSchedulerThread.Run`. I wanted to use a `PauseTokenSource` based on `AsyncManualResetEvent` found in `Microsoft.VisualStudio.Threading`. This library contains `Microsoft.VisualStudio.Threading.Analyzers` which helps resolve deadlocks. Due to the way warnings are configured as errors, I was forced to contend with each one which is how I ended up touching so many files. I attempted to resolve all material warnings without making breaking changes. There are still a couple places where async is waited synchronously or sync method is called in an async method, because I didn't want to change the signature of public methods.

I also used `JoinableTask` in shutdown methods where the intention is to wait for a task started elsewhere to finish. This might end up being the key to solving the shutdown deadlocks I was seeing in edge cases.

RELATED ISSUES:
- #2354
- #2133

### CODE REVIEW

File: Directory.Packages.props
Comment (lahma) [inline Directory.Packages.props:1]: This file looks like a merge/rebase gone wrong.

File: src/Quartz.Benchmark/DefaultThreadPoolBenchmark.cs [hunk 1]
```diff
@@ -25,15 +25,15 @@ public void RunInThread_CompletedTask_MaxConcurrencyIsMaxValue_SingleThreaded()
     }
 
     [Benchmark(OperationsPerInvoke = 1_000_000)]
-    public void RunInThread_CompletedTask_MaxConcurrencyIsMaxValue_MultiThreaded()
+    public async Task RunInThread_CompletedTask_MaxConcurrencyIsMaxValue_MultiThreaded()
```
Comment (lahma) [inline src/Quartz.Benchmark/DefaultThreadPoolBenchmark.cs:28]: maybe these quite safe changes for tests/benchmarks could be a separate PR (calling async versions)

File: src/Quartz.Server/Program.cs [hunk 2]
```diff
@@ -14,14 +15,16 @@ public static class Program
     /// <summary>
     /// Main.
     /// </summary>
-    public static void Main()
+    public static async Task Main()
```
Comment (lahma) [inline src/Quartz.Server/Program.cs:18]: Maybe also could be part of a separate cleanup PR as quite safe and outside of core domain

File: src/Quartz/Core/QuartzSchedulerThread.cs
Comment (lahma) [inline src/Quartz/Core/QuartzSchedulerThread.cs:1]: This is the scariest part and requires careful review and testing, crucial parts of Quartz

File: src/Quartz/Core/QuartzSchedulerThread.cs [hunk 3]
```diff
@@ -22,9 +22,10 @@
 using System.Data.Common;
 
 using Microsoft.Extensions.Logging;
-
+using Microsoft.VisualStudio.Threading;
```
Comment (lahma) [inline src/Quartz/Core/QuartzSchedulerThread.cs:25]: Core Quartz should not have this kind of dependencies

File: src/Quartz/Quartz.csproj [hunk 4]
```diff
@@ -38,6 +38,8 @@
     <PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" />
     <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" />
     <PackageReference Include="Microsoft.Extensions.Options" />
+    <PackageReference Include="Microsoft.VisualStudio.Threading" />
+    <PackageVersion Include="Microsoft.VisualStudio.Threading" Version="17.10.48" />
```
Comment (lahma) [inline src/Quartz/Quartz.csproj:42]: I'm uncomfortable with having this kind of dependency

**Review by lahma — COMMENTED** [review]
Had a quick read-through and far from proper review. Added some comments.

File: Directory.Packages.props
Comment (A9G-Data-Droid) [reply Directory.Packages.props:1]: Yes, that is likely.

File: src/Quartz.Benchmark/DefaultThreadPoolBenchmark.cs (hunk 1 above)
Comment (A9G-Data-Droid) [reply src/Quartz.Benchmark/DefaultThreadPoolBenchmark.cs:28]: Ok, if I do that I need to add more ignore clauses to stop the warnings from becoming errors and blocking build.

Comment (lahma) [reply src/Quartz.Benchmark/DefaultThreadPoolBenchmark.cs:28]: Well without the analyzer it wouldn't break the build? So my suggestions would be a separate PR for all test/example/benchmark code without analyzers so it should work the same.

File: src/Quartz/Core/QuartzSchedulerThread.cs (hunk 3 above)
Comment (A9G-Data-Droid) [reply src/Quartz/Core/QuartzSchedulerThread.cs:25]: Ok, that is the key to what I'm doing here. There are similar tools found in https://github.com/StephenCleary/AsyncEx that we could use instead. I chose to use the Microsoft implementation because I assume they are going to be more popular and better maintained over time. Are you okay with using the Stephen Cleary libraries instead?

Comment (lahma) [reply src/Quartz/Core/QuartzSchedulerThread.cs:25]: I'm still undecided. Because both are MIT licensed, we could just extract the minimal needed parts and add to Quartz codebase with attribution in each file's header...

Comment (A9G-Data-Droid) [reply src/Quartz/Core/QuartzSchedulerThread.cs:25]: Ok, I ended up doing that with the PauseToken anyway.

**Comment by A9G-Data-Droid** [comment]
Thank you for the review. This attempt turned out to be a "move fast and break things". I am going to step back and try again, paying attention to your feedback.
